### PR TITLE
CosmosDB Cache Invalidation Service

### PR DESCRIPTION
**Pull request by chetanmeh**
This PR implements a new service for performing Cache Invalidation when running with CosmosDB. See #4277 for details

## Description

To propagate changes across cluster this PR implements a new service which makes use of the [changefeedprocessor](https://github.com/Azure/azure-documentdb-changefeedprocessor-java) library to listen to changes happening in `whisks` and `subject` collection and then convert them into Kafka message events which can be sent to `cacheInvalidation`.

So each cluster would have its own cache invalidation service running which would ensure that changes done by other clusters are picked up and routed to local cluster `cacheInvalidation` topic

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#4277)

## My changes affect the following components

- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [ ] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [ ] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [ ] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by chetanmeh** [comment]
This PR is now ready for review

**Comment by chetanmeh** [comment]
With Azure/azure-documentdb-changefeedprocessor-java#61merged it should now be possible use the factory mode to pass the feed processor implementation. This would simplify our implementation where we need to use a singleton and set it up with all dependencies

**Comment by rabbah** [comment]
hi @chetanmeh thanks for your patience - started going through this just now.

File: common/scala/src/main/scala/org/apache/openwhisk/common/ExecutorCloser.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    try {
+      service.shutdown()
+      service.awaitTermination(timeout.toSeconds, TimeUnit.SECONDS)
+    } catch {
+      case e: InterruptedException =>
+        log.error("Error while shutting down the ExecutorService", e)
+        Thread.currentThread.interrupt()
+    } finally {
+      if (!service.isShutdown) {
+        log.warn("ExecutorService `{}` didn't shutdown property. Will be forced now.", service)
```
Comment (rabbah) [inline common/scala/src/main/scala/org/apache/openwhisk/common/ExecutorCloser.scala]: what's {} for?

File: core/cosmosdb/cache-invalidator/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# Copy app jars
+ADD build/distributions/cache-invalidator.tar /
+
+COPY init.sh /
+RUN chmod +x init.sh
+
+RUN adduser -D -u ${UID} -h /home/${NOT_ROOT_USER} -s /bin/bash ${NOT_ROOT_USER}
+
+# It is possible to run as non root if you dont need invoker capabilities out of the controller today
+#USER ${NOT_ROOT_USER}
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/Dockerfile]: is this relevant to this dockerfile?

File: core/cosmosdb/cache-invalidator/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+-->
+# OpenWhisk Cache Invalidator Service
+
+This service performs cache invalidation in OpenWhisk cluster to enable cache event propagation in multi region setups.
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: in OpenWhisk cluster -> in an OpenWhisk cluster?

File: core/cosmosdb/cache-invalidator/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+## Design
+
+OpenWhisk cluster uses a Kafka topic `cacheInvalidation` to communicate changes to any cached entity. Messages on this
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: An OpenWhisk cluster ...?

File: core/cosmosdb/cache-invalidator/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+topic are of the form
+
+```json
+{"instanceId":"controller0","key":{"mainId":"guest/hello"}}
+```
+
+When deploying multiple seprate cluster of OpenWhisk which do not share same Kafka instance we would need a way to
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: When deploying multiple seprate cluster of OpenWhisk which do not share same Kafka instance we would need a way to propagate the cache change event across cluster. ->  When deploying OpenWhisk across multiple nodes which do not share a common Kafka instance, we  need a way to propagate the cache-change events across the cluster.

File: core/cosmosdb/cache-invalidator/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+propagate the cache change event across cluster. For CosmosDB based setups this can be done by using [CosmosDB ChangeFeed][1]
+support. It enables reading changes that are made to any specific collection.
+
+This service makes use of [change feed processor java][2] library and listen to changes happening in `whisks` and `subject`
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: hyperlink to [2]? [change feed processor] Java library ...

File: core/cosmosdb/cache-invalidator/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+collection and then convert them into Kafka message events which can be sent to `cacheInvalidation` topic of local cluster
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: collection -> collections.

Comment (rabbah) [reply core/cosmosdb/cache-invalidator/README.md]: sent to `cacheInvalidation` topic of local cluster ->  sent to `cacheInvalidation` topic local to the cluster.

File: core/cosmosdb/cache-invalidator/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+The service needs following env variables to be set
+
+- `KAFKA_HOSTS` - For local env it can be set to `172.17.0.1:9093`. When using [OpenWhisk Devtools][3] based setup use `kafka`
+- `COSMOSDB_ENDPOINT` - Endpoint URL like https://<account>.documents.azure.com:443/
+- `COSMOSDB_KEY` - DB Key
+- `COSMOSDB_NAME` - DB name
+
+Upon startup it would create a collection to manage the lease data with name `cache-invalidator-lease`. For events sent by
+this service `instanceId` would be set to `cache-invalidator`
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: Events sent by this service `instanceId` are sent to the `cache-invalidator` topic.

File: core/cosmosdb/cache-invalidator/README.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+## Local Run
+
+Setup the OpenWhisk cluster using [devtools][3] but have it connect to CosmosDB. This would also start
+the [Kafka Topic UI][4] at port `8001`. Then if any change is made to db then you would see events set to Kafka topic like
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: Then when changes are made to the database, you should see events sent to the Kafka topic. For example, if a a package is created with `wsk package create test-package` using the `guest` account, the following event is generated:

File: core/cosmosdb/cache-invalidator/README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+For e.g. if any package is created like `wsk package create test-package` using `guest` account then following event can
+be seen
+
+```json
+ {"instanceId":"cache-invalidator","key":{"mainId":"guest/test-package"}}
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/README.md]: Should the event register the artifact type? We don't use it today since all assets are unique and share the cache... so I suppose I answered my own question and it's not necessary.

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/Main.scala [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+object Main {
+  def main(args: Array[String]): Unit = {
+    implicit val system: ActorSystem = ActorSystem("cache-invalidator-actor-system")
+    implicit val materializer: ActorMaterializer = ActorMaterializer()
+    ConfigMXBean.register()
+    Kamon.loadReportersFromConfig()
+    val invalidatorConfig = CacheInvalidatorConfig.getInvalidatorConfig()(system.settings.config)
+    val port = invalidatorConfig.port
+    //TODO HTTPS for ping/metric endpoint?
```
Comment (rabbah) [inline core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/Main.scala]: how do you check liveness without ping?
the basic http service should be exposing /ping already no?

File: common/scala/src/main/scala/org/apache/openwhisk/common/ExecutorCloser.scala (hunk 1 above)
Comment (chetanmeh) [reply common/scala/src/main/scala/org/apache/openwhisk/common/ExecutorCloser.scala]: Thats [slf4j placeholder format][1] usage. I can probably use Scala string interpolation support also

[1]: https://www.slf4j.org/faq.html#logging_performance

File: core/cosmosdb/cache-invalidator/Dockerfile (hunk 2 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/Dockerfile]: Yup per [best practices][1] its always better to use non root user

[1]: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user

File: core/cosmosdb/cache-invalidator/README.md (hunk 10 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/README.md]: As you mentioned it is not emitted by default logic and also not used. Later we may want to emit `entityType` if needed. Or if we start pushing subject invalidation events

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/Main.scala (hunk 11 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/Main.scala]: Yes its uses basic http service so ping is supported. The purpose of `TODO` here was to determine if we need to expose the basic service over HTTPS or not. If yes then that would require copying the required cert files etc.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/OpenWhisk-Playground/71/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/OpenWhisk-Playground/72/

**Comment by chetanmeh** [comment]
Rebased to latest master

**Comment by chetanmeh** [comment]
Currently the CosmosDB Change Feed Processor does not shutdown properly which causes stray thread to still run. This causes some log messages like

```
[2019-06-07T08:36:27.266Z] [ERROR] Other exception not handled happened: null 
java.lang.NullPointerException
	at com.microsoft.azure.documentdb.changefeedprocessor.services.CheckpointServices.getCheckpoint(CheckpointServices.java:77)
	at com.microsoft.azure.documentdb.changefeedprocessor.services.CheckpointServices.getCheckpointData(CheckpointServices.java:49)
	at com.microsoft.azure.documentdb.changefeedprocessor.services.ChangeFeedJob.QueryChangeFeed(ChangeFeedJob.java:159)
	at com.microsoft.azure.documentdb.changefeedprocessor.services.ChangeFeedJob.lambda$start$0(ChangeFeedJob.java:136)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
```

Opened Azure/azure-documentdb-changefeedprocessor-java#63 to clarify that aspect

[1]: https://scans.gradle.com/s/xmzxr775nasvo/tests/failed

File: common/scala/src/main/resources/logback.xml [hunk 12]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <configuration>
   <jmxConfigurator></jmxConfigurator>
+  <include optional="true" resource="whisk-logback.xml"/>
```
Comment (tysonnorris) [inline common/scala/src/main/resources/logback.xml]: I guess this is for local debugging and should not be checked in? If so, maybe add it to .gitignore and .dockerignore

File: core/cosmosdb/cache-invalidator/Dockerfile (hunk 2 above)
Comment (tysonnorris) [reply core/cosmosdb/cache-invalidator/Dockerfile]: Should NON_ROOT_USER be enabled here by default (not sure of any case it is required for cache-invalidator)

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidator.scala [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+  //CosmosDB changefeed support uses Java Logging.
+  // Those needs to be routed to Slf4j
+  SLF4JBridgeHandler.removeHandlersForRootLogger()
+  SLF4JBridgeHandler.install()
+
+  //TODO Replace with constant from RemoteCacheInvalidation
+  val cacheInvalidationTopic = "cacheInvalidation"
+
+  val instanceId = "cache-invalidator"
+  val whisksCollection = "whisks"
```
Comment (tysonnorris) [inline core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidator.scala]: Can these be taking from configuration?

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidatorConfig.scala [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+  }
+}
+
+case class ConnectionInfo(endpoint: String, key: String, db: String)
+
+case class FeedConfig(hostname: String, leaseCollection: String)
+
+case class InvalidatorConfig(port: Int, feedPublishTimeout: FiniteDuration, clusterId: Option[String])
+
+case class CacheInvalidatorConfig(globalConfig: Config) {
```
Comment (tysonnorris) [inline core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidatorConfig.scala]: Can we use pureconfig here to be consistent with other OW components?

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/KafkaEventProducer.scala [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.kafka.clients.producer.ProducerRecord
+
+import scala.collection.immutable.Seq
+import scala.concurrent.{ExecutionContext, Future, Promise}
+
+case class KafkaEventProducer(settings: ProducerSettings[String, String], topic: String)(
+  implicit system: ActorSystem,
+  materializer: ActorMaterializer)
+    extends EventProducer {
+  private val bufferSize = 100
```
Comment (tysonnorris) [inline core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/KafkaEventProducer.scala]: May need to make this configurable? I'm also not sure 100 is enough if kafka gets slowed down for any reason, but I'm not sure what the best safe minimum value would be.

File: common/scala/src/main/resources/logback.xml (hunk 12 above)
Comment (chetanmeh) [reply common/scala/src/main/resources/logback.xml]: Yes this is meant to allow other services based om common module to customize the logback config

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidatorConfig.scala (hunk 14 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidatorConfig.scala]: Yes. Purpose here is to adapt the config for test runs in a simplified manner. Our current approach in common and other module involve direct loading of global config which makes it very hard to manipulate the config for different test scenarios.

In this service I am trying to follow an approach where config is also passed down the line and hence not rely on a global singleton

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidator.scala (hunk 13 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/CacheInvalidator.scala]: Its fine for them being constant. `instanceId` is used as a marker name in the cache invalidation message sent to Kafka. There is no use of this instanceId apart from fact that controller would compare its own id with this and upon mismatch would invalidate the cache entry

```json
{"instanceId":"cache-invalidator","key":{"mainId":"guest/test-package"}}
```

`whisksCollection` is name of the collection from which change feed is run. This is always `whisks`. User would only be customizing the db name but collection name within db remains same. So fine to be constant

File: core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/KafkaEventProducer.scala (hunk 15 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/src/main/scala/org/apache/openwhisk/core/database/cosmosdb/cache/KafkaEventProducer.scala]: Makes sense. Made it configurable now

**Comment by chetanmeh** [comment]
Build is [passing][2] on [fork][1]

[1]: https://travis-ci.org/chetanmeh/incubator-openwhisk/builds/562915562
[2]: https://scans.gradle.com/s/qiq5y3mu7stxo/tests

File: core/cosmosdb/cache-invalidator/Dockerfile (hunk 2 above)
Comment (chetanmeh) [reply core/cosmosdb/cache-invalidator/Dockerfile]: Missed that. Enabled it now
